### PR TITLE
[WIP] Determine API root behind a proxy

### PR DESCRIPTION
**Pull request by manthey**
When behind a proxy, swagger doesn't work because of how we get the api url.  This now has a setting to override it.  If set to ".", the url will look ugly (as it has an extra . in the path, but will work both inside and outside of the proxy.  If set to "https?://(proxy host and port)/(proxy path)/api/v1", it will look nice and work through the proxy, but not locally.

A better technique would be to modify swagger-ui, but that seems heavy-handed for now.

I need to add tests for this and update the documentation, but wanted some feedback first.

This addresses issue #1054.

### CODE REVIEW

**Comment by manthey** [comment]
@zachmullen This addresses issue #1054.  If you would rather have a different setting name or think I should use a different solution, let me know.  If you like this, I'll update the documentation and add tests.

**Comment by zachmullen** [comment]
It seems we should add this logic into `getApiUrl`, or is there some reason not to?

Also, I thought the `tools.proxy.base` config setting would take care of this for us. Does it not?

**Comment by manthey** [comment]
`tools.proxy.base` doesn't fix the issue (I just checked).  Swagger doesn't reference things properly within the proxy (and it probably won't work both inside and outside of the proxy).

I can move the code to `getApiUrl`.

**Comment by manthey** [comment]
@zachmullen, Using `tools.proxy.base` *and* `tools.proxy.local` does work around the problem, but another technique would be to use the `Referer` header to determine the api root. (see current changes).  The virtue of this change is that (a) it reduces the amount of configuration that is needed, and (b) it works both inside and outside of the proxy.

File: girder/api/rest.py [hunk 1]
```diff
@@ -59,18 +59,29 @@ def getUrlParts(url=None):
...
     In a request thread, call this to get the path to the root of the REST API.
-    The returned path does *not* end in a forward slash.
+    The returned path does *not* end in a forward slash.  If no url is
+    specified, and there is a Referer header that includes the REST API root,
+    use that for determining the root.  This allows detection of the root from
+    within and outside of a proxy.
 
     :param url: URL from which to extract the base URL. If not specified, uses
-        `cherrypy.url()`
+        `cherrypy.url()` or the config value from ['server']['api_url'].
```
Comment (zachmullen) [inline girder/api/rest.py]: This doesn't seem to actually use the config value from ['server']['api_url'], am I missing something?

File: girder/api/rest.py [hunk 2]
```diff
@@ -59,18 +59,29 @@ def getUrlParts(url=None):
...
     """
+    apiPhrase = '/api/v1'
     url = url or cherrypy.url()
-    idx = url.find('/api/v1')
+    if url is None and 'Referer' in cherrypy.request.headers:
```
Comment (zachmullen) [inline girder/api/rest.py]: Will this block ever be entered? Wouldn't `cherrypy.url()` have to return `None` for that?

File: girder/api/rest.py (hunk 1 above)
Comment (manthey) [inline girder/api/rest.py]: My bad.  I removed the setting but didn't revert the text.

File: girder/api/rest.py (hunk 2 above)
Comment (manthey) [inline girder/api/rest.py]: Apparently I'm just paying attention this morning...

**Comment by zachmullen** [comment]
I'm not sure we should use the Referer header, for a couple of reasons:

1. In the normal case, the user agent sets the Referer header to the URL of the page that *referred* the user agent to the requested URL. So if someone makes a CORS request to our API, this would actually be set to the URL of the third-party site the user was looking at when the request was made.
2. In the malicious user case, the user can set the Referer header to whatever value they want, meaning someone could fool a romanesco task into uploading result data to any URL they wanted, for instance.

**Comment by zachmullen** [comment]
To be fair, it seems to me that it would also be possible for a malicious user spoof the hostname in an HTTP request, meaning cherrypy.url() could also be wrong. I'm wondering if we should just have a config value or setting for this rather than trying to derive this info from request parameters, which can almost all be spoofed.

**Comment by manthey** [comment]
If we care about header spoofing, then there is no safe way to determine the url both inside and outside of a proxy.  Inside of a proxy, the cherrypy.url() will be the internal url.  I think the only safe method that works on both sides of the proxy is using `.` as we used to do.

**Comment by zachmullen** [comment]
True, though that still leaves us with the problem of 1). The referer header will almost never contain what we want it to, and there's a small chance it could contain something that will mess things up, i.e. if a third-party site makes a request to girder from a page that happens to contain "/api/v1" for whatever reason.

I'm thinking we may just want to improve our documentation for the cherrypy proxy config values, or add our own absolute URL config value to get around this issue completely.

**Comment by zachmullen** [comment]
@manthey can we close this or do we have plans for it?

**Pull request closed without merging**